src/storage/src/source/kafka.rs

```diff
@@ -1549,23 +1549,68 @@ mod tests {
 }
 
 /// Fetches the list of partitions and their corresponding high watermark
-fn fetch_partition_info<C: ClientContext>(
-    client: &Client<C>,
+fn fetch_partition_info<C: ConsumerContext>(
+    consumer: &BaseConsumer<C>,
     topic: &str,
     fetch_timeout: Duration,
 ) -> Result<BTreeMap<PartitionId, WatermarkOffsets>, anyhow::Error> {
-    let pids = get_partitions(client, topic, fetch_timeout)?;
-
-    let mut result = BTreeMap::new();
+    let pids = get_partitions(consumer.client(), topic, fetch_timeout)?;
+
+    // If there is only a single partition, `fetch_watermarks` is more efficient than
+    // `offsets_for_times`. While both require two requests to the broker (one for each watermark),
+    // they are pipelined for `fetch_watermarks`.
+    // If there is more than one partition, we prefer using `offsets_for_times` because it requires
+    // fewer requests to the broker (2 instead of `pids.len() * 2`).
+    if pids.len() == 1 {
+        let pid = pids[0];
+        let (low, high) = consumer
+            .client()
+            .fetch_watermarks(topic, pid, fetch_timeout)?;
+        let watermarks = WatermarkOffsets {
+            low: low.try_into().expect("invalid negative offset"),
+            high: high.try_into().expect("invalid negative offset"),
+        };
+        return Ok([(pid, watermarks)].into());
+    }
```

Comment:
It's a bit of a shame to have to fetch the low water mark at all. We don't use it for anything important, AFAICT. cc @petrosagg to check me on that. Could we just ... not? It seems like it's only used to initialize the start offset of a partition, and that could just as well be set to 0.

Reply:
Ha, I was wondering about this! Would be great if we could remove it. Then we'd need fewer requests, and we could throw out the special case here and just always use `offsets_for_time`.

Reply:
Looked at the code and I agree that the low watermark could be removed. Let's try it and see what CI thinks of that